File: Algorithm_Exercises/Algorithm (4th edition)/Algorithm/Chapter4/Graph/BreadthFirstPaths.cs
using Chapter1;
using System;
using System.Collections.Generic;
using System.Runtime.ConstrainedExecution;
using System.Text;

namespace Chapter4
{
    class BreadthFirstPaths : AbsSearch
    {
        private bool[] marked;
        private int[] edgeTo;
        
        public BreadthFirstPaths(IGraph g,int s):base(g,s)
        {
            marked = new bool[g.V()];
            edgeTo = new int[g.V()];
            Bfs(g, s);
        }

        private void Bfs(IGraph g, int s)
        {
            var queue = new Chapter1.Queue<int>();
            marked[s] = true; //起点被标记
            queue.Enqueue(s);
            while(!queue.IsEmpty)
            {
                int v= queue.Dequeue(); //从队列中删除一个顶点
                foreach(int w in g.Adj(v))
                    if(!marked[w])
                    {
                        edgeTo[w] = v;
                        marked[w] = true;
                        queue.Enqueue(w); 
                    }
            }
        }

        public override int Count()
        {
            throw new NotImplementedException();
        }

        public override bool HasPathTo(int v)
        {
            return marked[v];
        }

        public override bool Marked(int v)
        {
            return marked[v];
        }

        public override IEnumerable<int> PathTo(int v)
        {
            if (!HasPathTo(v))
                return null;
            Chapter1.Stack<int> path = new Chapter1.Stack<int>();
            for (int x = v; x != S; x = edgeTo[x])
                path.Push(x);
            path.Push(S);
            return path;
        }
    }
}

File: Learn_DataStructure/Graph/BreadthFIrstPaths.java
package Graph;

import StackAndQueue.Queue;
import StackAndQueue.Stack;

/**
 * Created by sunkai on 2017/9/4
 */
public class BreadthFIrstPaths {
    private boolean[] marked;
    private int[] edgeTo;
    private final int s;

    public BreadthFIrstPaths(Graph G, int s) {
        marked = new boolean[G.V()];
        edgeTo = new int[G.V()];
        this.s = s;
        bfs(G, s);
    }

    private void bfs(Graph G, int s) {
        Queue<Integer> queue = new Queue<>();
        marked[s] = true;
        queue.enqueue(s);
        while (!queue.isEmpty()) {
            int v = queue.dequeue();
            for (int w : G.adj(v)) {
                if (!marked[w]) {
                    edgeTo[w] = v;
                    marked[w] = true;
                    queue.enqueue(w);
                }
            }
        }
    }

    public boolean hasPathTo(int v) {
        return marked[v];
    }

    public Iterable<Integer> pathTo(int v) {
        if (!hasPathTo(v))
            return null;
        Stack<Integer> path = new Stack<>();
        for (int x = v; x != s; x = edgeTo[x])
            path.push(x);
        return path;
    }
}